### PR TITLE
Fix param parsing.

### LINKED ISSUE #285: Wrong number of Parameter for numpy array.
```
In [1]: from numpydoc.docscrape import NumpyDocString
   ...: from numpy import array
   ...: len(NumpyDocString(array.__doc__)['Parameters'])
Out[1]: 1
```

(should be 6)

Again due to a dedent issue: 

```
$ git diff
diff --git a/numpydoc/docscrape.py b/numpydoc/docscrape.py
index d79992c..3db20fe 100644
--- a/numpydoc/docscrape.py
+++ b/numpydoc/docscrape.py
@@ -219,6 +219,7 @@ class NumpyDocString(Mapping):
                 yield name, self._strip(data[2:])

     def _parse_param_list(self, content, single_element_is_type=False):
+        content = dedent_lines(content)
         r = Reader(content)
         params = []
         while not r.eof():
```

Fixes it. 

Not quite sure why it is not found in tests...

### PR DESCRIPTION
**Pull request by Carreau**
Closes #285

This fixes two tings:

 - When first sentence of the docstring is onteh first line, Parameters
 is not properly parse, which for example mis parsed numpy.array
 docstring.

 - many project have paremeters description list with ` :` afer the
 name, even if no type is present. If there is no space after the `:`
 the parameter name includes the ` :` which is most likely wrong.

### CODE REVIEW

**Comment by Carreau** [comment]
> to get the dedenting correct for the entire docstring when the first line doesn't match the indenting scheme of the rest of the docstring?

That was the first fix I tried, but this make `validate` fails as well as other things which are attempting to tell you when the docstring is not valid or have extra spaces in place it does not like.

Personally I end up always calling it as so `Numpydoc(dedent_but_first(docstring))`, to work around some other inconsistencies.

**Comment by rossbar** [comment]
> That was the first fix I tried, but this make validate fails as well as other things which are attempting to tell you when the docstring is not valid or have extra spaces in place it does not like.

Yes, I noticed this too when I was messing around with alternatives (see #287, which will fail one validation test). It seems to me much more straightforward to catch issues due to dedenting errors as far upstream as possible, rather than modifying the downstream functions to handle the corner cases of less-robust docstring cleaning.

However, if there are strong backwards compatibility concerns with the validation module, then the approach in #287 is a non-starter.

**Comment by Carreau** [comment]
Yeah, I went with not validating, and just re-emitting the docstring from the Numpydoc `_parsed_data` data structure as the `__repr__` does also some weird things and ended up with a docstring reformatter.

**Comment by rossbar** [comment]
Hmm, on the one hand the validation should definitely catch indentation errors, but it would be nice if the docstring parsing were more resilient to indentation problems.

**Comment by Carreau** [comment]
Yeah, I think the two logics are intertwined too much, and the "parsing"/"guessing"/"validating" should be different steps. 

There are many docstrings around that have invalid section names, and right now can't be parsed as Numpydoc Errors immediately, though that might be for a longer term project to clean that up.

**Comment by rossbar** [comment]
> There are many docstrings around that have invalid section names, and right now can't be parsed as Numpydoc Errors immediately, though that might be for a longer term project to clean that up.

... including in numpy itself! (numpy/numpy#16791)

I agree it's probably a larger project, so let's not worry about it in this PR. I think the changes here look good. The only additions I'd make is to apply a similar `pytest.mark.parametrize` to the `test_returns` and `test_other_parameters` tests, since they also are affected by the changes to `_parse_param_list`. I also slightly prefer the parametrization I used in #287 as it's a little shorter, but YMMV.

**Comment by jnothman** [comment]
> many project have paremeters description list with : afer the name, even if no type is present. If there is no space after the : the parameter name includes the : which is most likely wrong.

The problem being that this matched the relatively obscure syntax for ReST definition lists...

**Comment by Carreau** [comment]
ok, I made `doc` a parameterized fixture so every test using it are now parametrized on `''` (flush) vs `'\n    '` (newline indented)

**Comment by Carreau** [comment]
Anything I can do to push this forward ?

File: numpydoc/tests/test_docscrape.py [hunk 1]
```diff
@@ -921,6 +926,20 @@ class BadSection:
...
+        Parameters
+        ----------
+
+        data :
+            some stuff, technically invalid
+        """)
+
+
+def test_trailing_colon():
+    assert doc8['Parameters'][0].name == 'data'
```
Comment (larsoner) [inline numpydoc/tests/test_docscrape.py:942]: PEP8 two newlines before new func (or after this one)
```suggestion
    assert doc8['Parameters'][0].name == 'data'

```

File: numpydoc/tests/test_docscrape.py [hunk 2]
```diff
@@ -921,6 +926,20 @@ class BadSection:
 def test_empty_first_line():
     assert doc7['Summary'][0].startswith('Doc starts')
 
+doc8 = NumpyDocString("""
+
+        Parameters wit colon and no types:
```
Comment (larsoner) [inline numpydoc/tests/test_docscrape.py:931]:
```suggestion
        Parameters with colon and no types:
```

**Comment by Carreau** [comment]
Thanks !

**Comment by larsoner** [comment]
Thanks @Carreau !
